scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/link-list-handler.js

```diff
@@ -0,0 +1,88 @@
+/**
+ * Check text nodes at the top level of the static_methods section.
+ * Text nodes here may only contain newlines.
+ */
+function checkTextNode(node, logError) {
+  const newlinesOnly = /^\n*$/;
+  if (!node.value.match(newlinesOnly)) {
+    logError(
+      "Text nodes in section may only contain newlines",
+      "text-nodes-in-section"
+    );
+  }
+}
+
+/**
+ * Check elements at the top level of the static_methods section.
+ * Only `dl` elements are allowed.
+ */
+function checkElementNode(node, logError) {
+  if (node.tagName !== "dl") {
+    logError(
+      "Section may only contain DL elements",
+      "non-dl-elements-in-section"
+    );
+    return;
+  }
+  checkDL(node, logError);
+}
+
+/**
+ * Check the structure of a `dl` element that represents a list of links.
+ */
+function checkDL(node, logError) {
+  let title = "";
+  for (const child of node.children) {
+    // We are only going to check `dt` elements, and will tolerate
+    // any others
+    if (child.tagName === "dt") {
+      if (child.children.length !== 1 || child.children[0].tagName !== "a") {
+        logError(
+          "DT elements in link lists must contain a single anchor element",
+          "link-list-dt-element-child-is-anchor"
+        );
+        return;
+      }
+      const link = child.children[0];
+      if (link.children.length !== 1 || link.children[0].tagName !== "code") {
+        logError(
+          "Anchor elements in link lists must contain a single code element",
+          "link-list-anchor-element-child-is-code"
+        );
+        return;
+      }
+      const code = link.children[0];
+      if (code.children.length !== 1 || code.children[0].type !== "text") {
+        logError(
+          "Code elements in link lists must contain a single text node",
+          "link-list-code-element-child-is-text"
+        );
+        return;
+      }
+      if (code.children[0].value.localeCompare(title, "en") <= 0) {
+        logError(
+          "Links in link lists must be listed in alphabetical order",
+          "link-list-alpha-order"
+        );
+      }
+      title = code.children[0].value;
+    }
+  }
+}
+
+function checkLinkList(section, logError) {
```

Comment:
I'm finding this function a little hard to follow. I guess we're trying to figure out if the section is well-formed, but this code is a bit shy about saying what that is (we check text nodes and elements, but for what? Nothing really says). I've inferred that a section should:

* Contain one `<h2>` element (which we've already cut off)
* Contain one or more `<dl>` elements, such that each `<dl>` element:
  * Contains zero or more `<dt>` elements in alphabetical order (based on their text content)
  * Each `<dt>`'s children are of the form `<a><code>textLiteral</code></a>`
* Contain only white space text literals
* Contain no other elements

Do I have that right? I think it might help if this were structured to show these requirements a bit more directly. It might be slightly more expensive, but, you could, for example:

1. Check section.children for forbidden element types and text literals
2. Visit each `<dl>`, collect the text of each `<dt>`, and check the alpha sort
3. Visit each `<dt>` and check its well-formedness.

Reply:
I've substantially rewritten this to be I hope more along the lines you have suggested here. I didn't use `visit` because to be honest I didn't think it would make this code any clearer.